### PR TITLE
[LADTable] Lad bounds listener FIX

### PR DESCRIPTION
**Pull request by jvigliotta**
Closes: #3113 

Fixes bounds issues, local time system formatter (as well as other missing formatters) and race condition logic edge cases.

EDIT:
Added support for local time system in generators.

## Author Checklist
Changes address original issue? Yes
Unit tests included and/or updated with changes? N/A
Command line build passes? Yes
Changes have been smoke-tested? Yes
Testing instructions included? Yes

### CODE REVIEW

File: example/generator/GeneratorMetadataProvider.js [hunk 1]
```diff
@@ -28,6 +28,15 @@ define([
                         domain: 2
                     }
                 },
+                {
+                    key: "local",
+                    name: "Time",
+                    format: "utc",
```
Comment (akhenry) [inline example/generator/GeneratorMetadataProvider.js:34]: Format should be `local`. That way it will show local time, instead of UTC.

File: src/plugins/LADTable/components/LADRow.vue [hunk 2]
```diff
@@ -64,14 +64,15 @@ export default {
...
-            return this.timestamp !== undefined ? this.formats[this.timestampKey].format(this.timestamp) : '---';
+            return this.timestamp !== undefined ? this.getFormattedTimestamp(this.timestamp) : '---';
         }
     },
     mounted() {
         this.metadata = this.openmct.telemetry.getMetadata(this.domainObject);
         this.formats = this.openmct.telemetry.getFormatMap(this.metadata);
         this.keyString = this.openmct.objects.makeKeyString(this.domainObject.identifier);
         this.bounds = this.openmct.time.bounds();
+        console.log('metadata, formats', this.metadata, this.formats);
```
Comment (akhenry) [inline src/plugins/LADTable/components/LADRow.vue:75]: Looks like some debug output has been left in?

File: src/plugins/LADTable/components/LADRow.vue [hunk 3]
```diff
@@ -89,6 +90,7 @@ export default {
         this.openmct.time.on('bounds', this.updateBounds);
 
         this.timestampKey = this.openmct.time.timeSystem().key;
+        console.log('timestamp key', this.timestampKey);
```
Comment (akhenry) [inline src/plugins/LADTable/components/LADRow.vue:93]: Debug output, should be removed.

File: src/plugins/LADTable/components/LADRow.vue [hunk 4]
```diff
@@ -110,24 +112,34 @@ export default {
...
+                    this.value = this.formats[this.valueKey].format(datum);
+                    limit = this.limitEvaluator.evaluate(datum, this.valueMetadata);
+                    if (limit) {
+                        this.valueClass = limit.cssClass;
+                    } else {
+                        this.valueClass = '';
+                    }
                 }
+            } else {
+                this.resetValues();
```
Comment (akhenry) [inline src/plugins/LADTable/components/LADRow.vue:130]: Let's reset values [when the user changes bounds](https://github.com/nasa/openmct/pull/3114/files#diff-283f9645a299cbc3994fbf01050dc9eeR161). I think that will simplify things.

File: example/generator/GeneratorMetadataProvider.js [hunk 5]
```diff
@@ -28,6 +28,15 @@ define([
                         domain: 2
                     }
                 },
+                {
+                    key: "local",
+                    name: "Time",
+                    format: "local-format",
```
Comment (jvigliotta) [inline example/generator/GeneratorMetadataProvider.js:34]: Changed to "local-format" as "local" wasn't formatting.

File: src/plugins/LADTable/components/LADRow.vue [hunk 6]
```diff
@@ -110,24 +112,34 @@ export default {
...
             }
         },
         shouldUpdate(newTimestamp) {
-            return (this.timestamp === undefined) ||
-                (this.inBounds(newTimestamp) &&
-                newTimestamp > this.timestamp);
+            let newTimestampInBounds = this.inBounds(newTimestamp),
+                noExistingTimestamp = this.timestamp === undefined,
+                existingTimestampInBounds = this.inBounds(this.timestamp),
```
Comment (akhenry) [inline src/plugins/LADTable/components/LADRow.vue:136]: Don't worry about whether the existing value is out of bounds. I think it will simplify things if we just always throw it away when the user changes bounds.

File: src/plugins/LADTable/components/LADRow.vue [hunk 7]
```diff
@@ -110,24 +112,34 @@ export default {
...
+                newTimestampIsLatest = newTimestamp > this.timestamp;
+
+            return newTimestampInBounds &&
```
Comment (akhenry) [inline src/plugins/LADTable/components/LADRow.vue:139]: I found the logic here kinda gnarly (though you've done a good job of explaining it using sensible variable names). I think we can simplify it if we don't consider the whether the existing values are still in bounds.

File: src/plugins/LADTable/components/LADRow.vue [hunk 8]
```diff
@@ -110,24 +112,34 @@ export default {
...
-                if (limit) {
-                    this.valueClass = limit.cssClass;
-                } else {
-                    this.valueClass = '';
+            if(datum) {
+                let newTimestamp = this.getParsedTimestamp(datum),
+                    limit;
+
+                if(this.shouldUpdate(newTimestamp)) {
+                    this.timestamp = this.getParsedTimestamp(datum);
```
Comment (akhenry) [inline src/plugins/LADTable/components/LADRow.vue:120]: I think we can just use `newTimestamp` here, instead of calling the function again? `updateValues` will be called with very high frequency, we need to be very wary of performance here.

File: example/generator/GeneratorMetadataProvider.js (hunk 5 above)
Comment (akhenry) [reply example/generator/GeneratorMetadataProvider.js:34]: Oh, yeah, I made a comment about this and then deleted it because I realized that the local formatter is only available when the local time system is installed, which it's not by default 😄 So we'll have to stick with `utc` for now.

Comment (akhenry) [reply example/generator/GeneratorMetadataProvider.js:34]: Which, actually, makes me think we just shouldn't have `local` values at all, because they will be shown in the UI. The best thing to do here might be to set it to use `local-format`, but comment out the metadata for `local`. Make a note about needing to enable the local time system to make use of it.

It's the one time I'll allow commented code to be checked in 😄

File: src/plugins/LADTable/components/LADRow.vue (hunk 2 above)
Comment (jvigliotta) [reply src/plugins/LADTable/components/LADRow.vue:75]: Removed

File: src/plugins/LADTable/components/LADRow.vue (hunk 3 above)
Comment (jvigliotta) [reply src/plugins/LADTable/components/LADRow.vue:93]: Removed
